### PR TITLE
Exercise3

### PR DESCRIPTION
**Pull request by MatzeK105**
Finished exercise 3:
- restructured repo
- setup graphql, apollo
- reconfigured eslint, jest
- updated CI
- implemented listing all posts, creating, upvoting, downvoting, deleting a post

### CODE REVIEW

File: .github/workflows/node.js.yml [hunk 1]
```diff
@@ -29,9 +29,16 @@ jobs:
...
-      working-directory: webapp/sdf
+      working-directory: webapp/
     - run: npm run lint 
-      working-directory: webapp/sdf
+      working-directory: webapp/
     - run: npm run test:unit
-      working-directory: webapp/sdf
+      working-directory: webapp/
+    - run: npm ci
+      working-directory: backend/
```
Comment (steven-mi) [inline .github/workflows/node.js.yml:38]: I highly suggest to create a new workflow for each part of your code. While it works putting all in one, it becomes very messy when the workflow gets too large

File: .github/workflows/node.js.yml [hunk 2]
```diff
@@ -29,9 +29,16 @@ jobs:
...
+    - run: npm run lint
+      working-directory: backend/
+    - run: npm run test:unit
```
Comment (steven-mi) [inline .github/workflows/node.js.yml:41]: ⭐ For the apollo-server-testing setup and its integration in the CI/CD pipeline.

File: backend/src/resolvers.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  //       users.splice(index, 1);
+  //       return true;
+  //     }
+  //   }
+  // };
+
+  export const resolvers = {
+    Query: {
+      posts: (parent, args, context) => context.dataSources.pdb.allPosts(),
+      users: (parent, args, context) => context.dataSources.udb.allUsers()
```
Comment (steven-mi) [inline backend/src/resolvers.js:4]: ⭐ For listing all posts
⭐ For listing all users

File: backend/src/resolvers.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    },
+    Mutation: {
+      write: (parent, args, context) => context.dataSources.pdb.createPost(args.post.title, args.post.author.id),
+      upvote: (parent, args, context) => context.dataSources.pdb.votePost(args.id, args.voter.id, 1),
+      downvote: (parent, args, context) => context.dataSources.pdb.votePost(args.id, args.voter.id, -1),
```
Comment (steven-mi) [inline backend/src/resolvers.js:9]: ⭐ For creating a post
⭐ For upvoting a post

File: backend/src/resolvers.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      delete: (parent, args, context) => context.dataSources.pdb.deletePost(args.id)
+    },
+    Post:{
+      author: (parent, args, context) => context.dataSources.udb.getUserById(parent.user_id),
+      votes: (parent, args, context) => context.dataSources.pdb.getVotes(parent.id)
+    },
+    User: {
+      posts: (parent, args, context) => context.dataSources.pdb.getAllPostsFromUser(parent.id)
```
Comment (steven-mi) [inline backend/src/resolvers.js:17]: ⭐ For 'indefinitely' nestable queries.

File: backend/src/datasource/post-datasource.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    votePost(post_id, user_id, value) {
+        let post = this.posts.find(post => post.id == parseInt(post_id));
+        
+        if (post) {
+            let vote = post.votes.find(vote => vote.user_id == user_id)
+
+            if (!vote) {
+                post.votes.push({ user_id: user_id, value: value });
+                this.posts = this.posts.map((local_post) => (local_post.post_id == post_id) ? post : local_post)
+            }
```
Comment (steven-mi) [inline backend/src/datasource/post-datasource.js:57]: ⭐ For preventing users to upvote the same post multiple times.

File: backend/src/datasource/post-datasource.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            else {
+                vote.value = value
+                post.votes[post.votes.indexOf(vote)] = vote
+                this.posts[this.posts.indexOf(post)] = post
+            }
```
Comment (steven-mi) [inline backend/src/datasource/post-datasource.js:62]: I don't really understand this. It seems very redundant. If the user didn't vote for this post then you are overwriting the vote with the index of the vote from line 52 with the vote you got from line 52

File: backend/src/typeDefs.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+//   }
+
+//   type Query {
+//     findUser(id: ID!): User
+//   }
+
+//   type Mutation {
+//     deleteUser(id: ID!): Boolean
+//   }
+// `;
```
Comment (steven-mi) [inline backend/src/typeDefs.js]:
```suggestion
```
Not needed

File: backend/src/typeDefs.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+type Query {
+  posts: [Post]
+  users: [User]
+}
+
+type Mutation {
+  write(post: PostInput!): Post
+  # 🚀 OPTIONAL
```
Comment (steven-mi) [inline backend/src/typeDefs.js]:
```suggestion
```
Emoji in code not good

File: backend/src/typeDefs.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  delete(id: ID!): Post
+
+  # ⚠️ FIXME in exercise #4
+  # mock voter until we have authentication
```
Comment (steven-mi) [inline backend/src/typeDefs.js]:
```suggestion
```
Emoji in code not good

File: backend/src/typeDefs.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  upvote(id: ID!, voter: UserInput!): Post
+
+  # 🚀 OPTIONAL
+  downvote(id: ID!, voter: UserInput!): Post
+}
+
+input PostInput {
```
Comment (steven-mi) [inline backend/src/typeDefs.js:33]:
```suggestion
```
Emoji in code not good

File: backend/src/typeDefs.js [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  title: String!
+
+  # ⚠️ FIXME in exercise #4
+  # mock author until we have authentication
```
Comment (steven-mi) [inline backend/src/typeDefs.js]:
```suggestion
```
Emoji in code not good

File: backend/src/resolvers.js [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+  //   Mutation: {
+  //     deleteUser: (parent, { id }) => {
+  //       const index = users.findIndex(user => user.id === id);
+  //       if (index < 0) return false;
+  //       users.splice(index, 1);
+  //       return true;
+  //     }
+  //   }
+  // };
+
```
Comment (steven-mi) [inline backend/src/resolvers.js]:
```suggestion
```
Not needed

**Review by steven-mi — APPROVED** [review]
Review from @Systems-Development-and-Frameworks/abteilung6:

You get 9/10 ⭐. The last ⭐ is given, after you reviewed a pull request. Below is a more detailed review:

⭐ For correct folder structure, basic apollo server setup and it's linter in the CI/CD pipeline.

- Yes! `backend` folder has a `src` folder with all source files.

⭐ For implementing a custom data source for apollo-server
- Yes! Custom data sources are available at: `backend/src/datasource`

File: backend/src/datasource/post-datasource.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
+const {DataSource} = require('apollo-datasource');
+
+export class PostDatasource extends DataSource {
```
Comment (medizinmensch) [inline backend/src/datasource/post-datasource.js:3]: ⭐ For implementing a custom data source for apollo-server

**Review by medizinmensch — APPROVED** [review]
> 
> 
> Review from @Systems-Development-and-Frameworks/abteilung6:
> 
> You get 9/10 ⭐. The last ⭐ is given, after you reviewed a pull request. Below is a more detailed review:
> 
> ⭐ For correct folder structure, basic apollo server setup and it's linter in the CI/CD pipeline.
> 
>     * Yes! `backend` folder has a `src` folder with all source files.
> 
> 
> ⭐ For implementing a custom data source for apollo-server
> 
>     * Yes! Custom data sources are available at: `backend/src/datasource`

A huge thanks to @steven-mi !! Very helpful, detailed and concise review! Wow!

I only added the comment for the custom data source (even though @steven-mi already rightfully gave points for that).

So you're at 9/10 :star:

Did you review another groups code?

File: backend/src/datasource/post-datasource.js (hunk 14 above)
Comment (MatzeK105) [reply backend/src/datasource/post-datasource.js:3]: I started reviewing Abteilung6 yesterday. But I gave one comment and then had to interrupt my work. When I returned I noticed that Robert already reviewed the PR. I had nothing to add to Roberts review. You can find my comment here (https://github.com/Systems-Development-and-Frameworks/abteilung6/pull/5#discussion_r529623107)

File: .github/workflows/node.js.yml (hunk 1 above)
Comment (MatzeK105) [reply .github/workflows/node.js.yml:38]: Thanks we will take this into account.

**Pull request merged**